src/Compilers/CSharp/Test/Syntax/Parsing/ReadOnlyStructs.cs

```diff
@@ -0,0 +1,173 @@
+            );
+        }
+
+        [Fact]
+        public void ReadOnlyClassErr()
+        {
+            var text = @"
+class Program
+{
+    readonly class S1{}
+
+    public readonly delegate ref readonly int S2();
+
+    readonly public interface S3{}
+}
+";
+
+            var comp = CreateCompilationWithMscorlib45(text, parseOptions: TestOptions.Regular.WithLanguageVersion(LanguageVersion.Latest), options: TestOptions.DebugDll);
+            comp.VerifyDiagnostics(
+                // (4,20): error CS0106: The modifier 'readonly' is not valid for this item
+                //     readonly class S1{}
+                Diagnostic(ErrorCode.ERR_BadMemberFlag, "S1").WithArguments("readonly").WithLocation(4, 20),
+                // (6,47): error CS0106: The modifier 'readonly' is not valid for this item
+                //     public readonly delegate ref readonly int S2();
+                Diagnostic(ErrorCode.ERR_BadMemberFlag, "S2").WithArguments("readonly").WithLocation(6, 47),
+                // (8,31): error CS0106: The modifier 'readonly' is not valid for this item
+                //     readonly public interface S3{}
+                Diagnostic(ErrorCode.ERR_BadMemberFlag, "S3").WithArguments("readonly").WithLocation(8, 31)
+            );
+        }
+
+        [Fact]
+        public void ReadOnlyRefStruct()
+        {
+            var text = @"
+class Program
+{
+    readonly ref struct S1{}
+
+    unsafe readonly public ref struct S2{}
+}
+";
+
+            var comp = CreateCompilationWithMscorlib45(text, parseOptions: TestOptions.Regular.WithLanguageVersion(LanguageVersion.Latest), options: TestOptions.DebugDll.WithAllowUnsafe(true));
+            comp.VerifyDiagnostics(
+            );
+        }
+
+        [Fact]
+        public void ReadOnlyStructPartialMatchingModifiers()
+        {
+            var text = @"
+class Program
+{
+    readonly partial struct S1{}
+
+    readonly partial struct S1{}
+
+    readonly partial ref struct S2{}
+
+    readonly partial ref struct S2{}
+}
+";
+
+            var comp = CreateCompilationWithMscorlib45(text, parseOptions: TestOptions.Regular.WithLanguageVersion(LanguageVersion.Latest), options: TestOptions.DebugDll);
+            comp.VerifyDiagnostics(
+            );
+
+            var s1 = comp.GetTypeByMetadataName("Program+S1");
+            Assert.False(s1.IsByRefLikeType);
+            Assert.True(s1.IsReadOnly);
+
+            var s2 = comp.GetTypeByMetadataName("Program+S2");
+            Assert.True(s2.IsByRefLikeType);
+            Assert.True(s2.IsReadOnly);
+        }
+
+        [WorkItem(19808, "https://github.com/dotnet/roslyn/issues/19808")]
+        [Fact]
+        public void ReadOnlyStructPartialNotMatchingModifiers()
+        {
+            var text = @"
+class Program
+{
+    readonly partial struct S1{}
+
+    readonly partial ref struct S1{}
```

Comment:
how a modifier can be removed? 

---
In reply to: [118787804](https://github.com/dotnet/roslyn/pull/19713#discussion_r118787804) [](ancestors = 118787804)